src/DataProtection/DataProtection/src/Managed/AesGcmAuthenticatedEncryptor.cs

```diff
@@ -0,0 +1,231 @@
+                        aes.Decrypt(nonce, encrypted, tag, plaintext);
+                        return plaintext;
+                    }
+                    finally
+                    {
+                        // delete since these contain secret material
+                        Array.Clear(decryptedKdk, 0, decryptedKdk.Length);
+                        Array.Clear(derivedKey, 0, derivedKey.Length);
+                    }
+                }
+            }
+            catch (Exception ex) when (ex.RequiresHomogenization())
+            {
+                // Homogenize all exceptions to CryptographicException.
+                throw Error.CryptCommon_GenericError(ex);
+            }
+        }
+
+        public byte[] Encrypt(ArraySegment<byte> plaintext, ArraySegment<byte> additionalAuthenticatedData, uint preBufferSize, uint postBufferSize)
+        {
+            plaintext.Validate();
+            additionalAuthenticatedData.Validate();
+
+            try
+            {
+                // Allocate a buffer to hold the key modifier, nonce, encrypted data, and tag.
+                // In GCM, the encrypted output will be the same length as the plaintext input.
+                var retVal = new byte[checked(preBufferSize + KEY_MODIFIER_SIZE_IN_BYTES + NONCE_SIZE_IN_BYTES + plaintext.Count + TAG_SIZE_IN_BYTES + postBufferSize)];
+                int keyModifierOffset; // position in ciphertext.Array where key modifier begins
+                int nonceOffset; // position in ciphertext.Array where key modifier ends / nonce begins
+                int encryptedDataOffset; // position in ciphertext.Array where nonce ends / encryptedData begins
+                int tagOffset; // position in ciphertext.Array where encrypted data ends
+
+                checked
+                {
+                    keyModifierOffset = plaintext.Offset + (int)preBufferSize;
+                    nonceOffset = keyModifierOffset + KEY_MODIFIER_SIZE_IN_BYTES;
+                    encryptedDataOffset = nonceOffset + NONCE_SIZE_IN_BYTES;
+                    tagOffset = encryptedDataOffset + plaintext.Count;
+                }
+
+                // Randomly generate the key modifier and nonce
+                var keyModifier = _genRandom.GenRandom(KEY_MODIFIER_SIZE_IN_BYTES);
+                var nonceBytes = _genRandom.GenRandom(NONCE_SIZE_IN_BYTES);
+
+                Buffer.BlockCopy(keyModifier, 0, retVal, (int)preBufferSize, keyModifier.Length);
+                Buffer.BlockCopy(nonceBytes, 0, retVal, (int)preBufferSize + keyModifier.Length, nonceBytes.Length);
+
+                // At this point, retVal := { preBuffer | keyModifier | nonce | _____ | _____ | postBuffer }
+
+                // Use the KDF to generate a new symmetric block cipher key
+                // We'll need a temporary buffer to hold the symmetric encryption subkey
+                var decryptedKdk = new byte[_keyDerivationKey.Length];
```

Comment:
@GrabYourPitchforks would it be unwise to use the shared ArrayPool here?

Reply:
Going to merge this as is, can make this switch in the follow up / spanification PR if needed

Reply:
Correct, because it's a cryptographic secret. Within the runtime, our crypto libraries use an internal [`CryptoPool`](https://github.com/dotnet/runtime/blob/release/5.0/src/libraries/Common/src/System/Security/Cryptography/CryptoPool.cs) type instead of the public singleton `ArrayPool<T>.Shared` instance.

During the spanification work, this can be switched to using _stackalloc_ (the common case) or the pinned object heap. Then no pooling will be needed at all.